File: btd700/headset.py
"""
Qualcomm GAIA Protocol client for Sennheiser Bluetooth Headphones (HDB 630, Momentum 4, Accentum, etc.).
Allows controlling Active Noise Cancellation (ANC), Transparency mode, and Bass Boost over Bluetooth RFCOMM.
"""

import os
import re
import json
import time
import socket
import select
import subprocess
import threading
from typing import Optional, Tuple, List, Dict, Any

# Qualcomm GAIA V3 Constants for Sennheiser
GAIA_MAGIC = b"\xFF\x03"
VENDOR_SENNHEISER = 0x0495

CMD_REGISTER_NOTIFICATION = 0x0007
CMD_SET_BASS_BOOST = 0x1008
CMD_GET_BASS_BOOST = 0x1009
CMD_SET_TRANSPARENCY = 0x1A02
CMD_GET_TRANSPARENCY = 0x1A03
CMD_SET_ANC_STATUS = 0x1A04
CMD_GET_ANC_STATUS = 0x1A05
CMD_SET_ADAPTIVE_ANC = 0x1A06
CMD_GET_ADAPTIVE_ANC = 0x1A07
CMD_SET_ANTI_WIND = 0x1A08
CMD_GET_ANTI_WIND = 0x1A09

FEATURE_USER_EQ = 8
FEATURE_TRANSPARENCY = 12
FEATURE_ANC = 13

COMMON_RFCOMM_CHANNELS = [2, 1, 14, 15, 12, 3, 4, 5, 6, 7, 8, 9, 10, 11]

CONFIG_DIR = os.path.expanduser("~/.config/btd700")
CONFIG_FILE = os.path.join(CONFIG_DIR, "config.json")


def load_config() -> Dict[str, Any]:
    if os.path.exists(CONFIG_FILE):
        try:
            with open(CONFIG_FILE, "r") as f:
                return json.load(f)
        except Exception:
            pass
    return {}


def save_config(cfg: Dict[str, Any]):
    os.makedirs(CONFIG_DIR, exist_ok=True)
    try:
        with open(CONFIG_FILE, "w") as f:
            json.dump(cfg, f, indent=2)
    except Exception:
        pass


def find_paired_headsets() -> List[Dict[str, str]]:
    """Scan for paired Sennheiser headphones via bluetoothctl or system tools."""
    results = []
    # 1. Try bluetoothctl devices
    try:
        proc = subprocess.run(
            ["bluetoothctl", "devices"],
            capture_output=True,
            text=True,
            timeout=2.0
        )
        if proc.returncode == 0:
            for line in proc.stdout.splitlines():
                # Format: Device XX:XX:XX:XX:XX:XX Name
                parts = line.strip().split(" ", 2)
                if len(parts) >= 3 and parts[0] == "Device":
                    mac = parts[1].upper()
                    name = parts[2]
                    # Check for Sennheiser or known headphone keywords
                    lower = name.lower()
                    is_match = any(k in lower for k in ["hdb", "630", "sennheiser", "momentum", "accentum", "hd 450", "hd 350", "epos"])
                    results.append({
                        "mac": mac,
                        "name": name,
                        "recommended": is_match
                    })
    except Exception:
        pass

    # Sort so recommended Sennheiser devices come first
    results.sort(key=lambda x: not x["recommended"])
    return results


def get_default_mac() -> Optional[str]:
    """Retrieve saved headset MAC or auto-detect from paired Bluetooth devices."""
    cfg = load_config()
    if cfg.get("headset_mac"):
        return cfg["headset_mac"].upper()

    devices = find_paired_headsets()
    for d in devices:
        if d.get("recommended"):
            return d["mac"]
    if devices:
        return devices[0]["mac"]
    return None


class GaiaProtocol:
    @staticmethod
    def build(command: int, payload: bytes = b"", vendor: int = VENDOR_SENNHEISER) -> bytes:
        return (
            GAIA_MAGIC
            + len(payload).to_bytes(2, "big")
            + vendor.to_bytes(2, "big")
            + command.to_bytes(2, "big")
            + payload
        )

    @staticmethod
    def expected_response(command: int) -> int:
        return command | 0x0100

    @staticmethod
    def parse_many(buffer: bytes) -> Tuple[List[Tuple[int, int, bytes]], bytes]:
        packets = []
        offset = 0
        while len(buffer) - offset >= 8:
            if buffer[offset:offset + 2] != GAIA_MAGIC:
                next_magic = buffer.find(GAIA_MAGIC, offset + 1)
                if next_magic == -1:
                    return packets, buffer[-1:]
                offset = next_magic
                continue
            payload_len = int.from_bytes(buffer[offset + 2:offset + 4], "big")
            total = 8 + payload_len
            if len(buffer) - offset < total:
                break
            vendor = int.from_bytes(buffer[offset + 4:offset + 6], "big")
            command = int.from_bytes(buffer[offset + 6:offset + 8], "big")
            payload = buffer[offset + 8:offset + total]
            packets.append((vendor, command, payload))
            offset += total
        return packets, buffer[offset:]


class SennheiserHeadset:
    """Controls Active Noise Cancellation (ANC), Transparency, and EQ on Sennheiser headsets."""

    def __init__(self, mac: Optional[str] = None):
        self.mac = (mac or get_default_mac() or "").upper()
        self.sock: Optional[socket.socket] = None
        self.channel: Optional[int] = None
        self.buffer = b""
        self.lock = threading.Lock()

    def is_connected(self) -> bool:
        return self.sock is not None

    def close(self):
        if self.sock is not None:
            try:
                self.sock.close()
            except OSError:
                pass
            self.sock = None
        self.channel = None
        self.buffer = b""

    def __enter__(self):
        self.connect()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()

    def _connect_channel(self, channel: int, timeout: float = 3.0):
        self.close()
        sock = socket.socket(socket.AF_BLUETOOTH, socket.SOCK_STREAM, socket.BTPROTO_RFCOMM)
        sock.settimeout(timeout)
        sock.connect((self.mac, channel))
        self.sock = sock
        self.channel = channel

    def _recv_until(self, wanted_commands: set, timeout: float = 3.0) -> Tuple[int, int, bytes]:
        if self.sock is None:
            raise RuntimeError("Not connected to headset")
        deadline = time.monotonic() + timeout
        while time.monotonic() < deadline:
            packets, self.buffer = GaiaProtocol.parse_many(self.buffer)
            for vendor, command, payload in packets:
                if command in wanted_commands:
                    return vendor, command, payload
            remaining = max(0.1, deadline - time.monotonic())
            self.sock.settimeout(remaining)
            try:
                chunk = self.sock.recv(4096)
                if not chunk:
                    raise ConnectionError("RFCOMM channel closed")
                self.buffer += chunk
            except (socket.timeout, TimeoutError):
                pass
        raise TimeoutError(f"Timed out waiting for response to {wanted_commands}")

    def _exchange(self, command: int, payload: bytes = b"", timeout: float = 3.0) -> Tuple[int, int, bytes]:
        if self.sock is None:
            raise RuntimeError("Not connected to headset")
        packet = GaiaProtocol.build(command, payload)
        with self.lock:
            self.sock.sendall(packet)
            wanted = {GaiaProtocol.expected_response(command), command | 0x0180}
            return self._recv_until(wanted, timeout=timeout)

    def connect(self, timeout: float = 2.5) -> int:
        """Connect to the headset over RFCOMM, searching standard GAIA channels."""
        if not self.mac:
            raise ValueError("No Bluetooth MAC address provided. Specify MAC or pair headset with Linux.")

        # Check if cached channel in config
        cfg = load_config()
        cached_channel = cfg.get(f"channel_{self.mac}")
        channels_to_try = COMMON_RFCOMM_CHANNELS
        if cached_channel and cached_channel in channels_to_try:
            channels_to_try = [cached_channel] + [c for c in channels_to_try if c != cached_channel]

        errors = []
        for ch in channels_to_try:
            try:
                self._connect_channel(ch, timeout=timeout)
                vendor, command, payload = self._exchange(CMD_GET_ANC_STATUS, timeout=1.8)
                if vendor == VENDOR_SENNHEISER and command == GaiaProtocol.expected_response(CMD_GET_ANC_STATUS):
                    # Cache successful channel & MAC
                    cfg["headset_mac"] = self.mac
                    cfg[f"channel_{self.mac}"] = ch
                    save_config(cfg)
                    return ch
            except Exception as e:
                errors.append(f"ch {ch}: {e}")
                self.close()

        raise ConnectionError(
            f"Failed to connect to Sennheiser headset ({self.mac}) over RFCOMM. "
            f"Ensure the headset is powered on, paired, and within range.\nDetails: {', '.join(errors[:3])}"
        )

    def get_anc_status(self) -> bool:
        """Query if Active Noise Cancellation is currently enabled."""
        vendor, command, payload = self._exchange(CMD_GET_ANC_STATUS)
        if vendor == VENDOR_SENNHEISER and command == GaiaProtocol.expected_response(CMD_GET_ANC_STATUS) and payload:
            return bool(payload[0])
        return False

    def set_anc_status(self, enabled: bool) -> bool:
        """Turn Active Noise Cancellation ON (True) or OFF (False)."""
        payload = bytes([1 if enabled else 0])
        vendor, command, resp_payload = self._exchange(CMD_SET_ANC_STATUS, payload)
        return vendor == VENDOR_SENNHEISER

    def toggle_anc(self) -> bool:
        """Toggle ANC between enabled and disabled, returning new state."""
        current = self.get_anc_status()
        new_state = not current
        self.set_anc_status(new_state)
        return new_state

    def get_transparency(self) -> int:
        """Get transparency level (0-100%)."""
        vendor, command, payload = self._exchange(CMD_GET_TRANSPARENCY)
        if vendor == VENDOR_SENNHEISER and payload:
            return int(payload[0])
        return 0

    def set_transparency(self, level: int) -> bool:
        """Set transparency level (0-100%)."""
        level = max(0, min(100, int(level)))
        vendor, command, payload = self._exchange(CMD_SET_TRANSPARENCY, bytes([level]))
        return vendor == VENDOR_SENNHEISER

    def get_bass_boost(self) -> bool:
        """Query if Bass Boost is enabled."""
        vendor, command, payload = self._exchange(CMD_GET_BASS_BOOST)
        if vendor == VENDOR_SENNHEISER and payload:
            return bool(payload[0])
        return False

    def set_bass_boost(self, enabled: bool) -> bool:
        """Set Bass Boost ON or OFF."""
        vendor, command, payload = self._exchange(CMD_SET_BASS_BOOST, bytes([1 if enabled else 0]))
        return vendor == VENDOR_SENNHEISER

    def get_anc_strength(self) -> int:
        """Get ANC strength percentage (0-100%). 100% is maximum noise cancellation."""
        if not self.get_anc_status():
            return 0
        return 100 - self.get_transparency()

    def set_anc_strength(self, strength: int) -> bool:
        """Set ANC strength percentage (0-100%). 100% = max noise cancellation, 0% = max transparency."""
        strength = max(0, min(100, int(strength)))
        if strength > 0:
            self.set_anc_status(True)
        transparency = 100 - strength
        return self.set_transparency(transparency)

    def get_adaptive_anc(self) -> bool:
        """Query if Adaptive ANC is enabled."""
        try:
            vendor, command, payload = self._exchange(CMD_GET_ADAPTIVE_ANC, timeout=1.5)
            if vendor == VENDOR_SENNHEISER and payload:
                return bool(payload[0])
        except Exception:
            pass
        return False

    def set_adaptive_anc(self, enabled: bool) -> bool:
        """Turn Adaptive ANC ON (auto adjusts) or OFF (fixed level)."""
        try:
            vendor, command, payload = self._exchange(CMD_SET_ADAPTIVE_ANC, bytes([1 if enabled else 0]), timeout=1.5)
            return vendor == VENDOR_SENNHEISER
        except Exception:
            return False

    def get_anti_wind(self) -> str:
        """Get anti-wind reduction mode ('off', 'auto', 'max')."""
        try:
            vendor, command, payload = self._exchange(CMD_GET_ANTI_WIND, timeout=1.5)
            if vendor == VENDOR_SENNHEISER and payload:
                val = payload[0]
                return {0: "off", 1: "auto", 2: "max"}.get(val, "off")
        except Exception:
            pass
        return "off"

    def set_anti_wind(self, mode: str) -> bool:
        """Set anti-wind reduction mode ('off', 'auto', 'max')."""
        m_map = {"off": 0, "auto": 1, "max": 2}
        val = m_map.get(mode.lower(), 0)
        try:
            vendor, command, payload = self._exchange(CMD_SET_ANTI_WIND, bytes([val]), timeout=1.5)
            return vendor == VENDOR_SENNHEISER
        except Exception:
            return False

    def get_state(self) -> Dict[str, Any]:
        """Fetch complete headset state (ANC, Strength, Transparency, Bass Boost, Wind)."""
        anc_on = self.get_anc_status()
        trans = self.get_transparency()
        return {
            "mac": self.mac,
            "connected": self.is_connected(),
            "anc_enabled": anc_on,
            "anc_strength": (100 - trans) if anc_on else 0,
            "transparency": trans,
            "adaptive_anc": self.get_adaptive_anc(),
            "anti_wind": self.get_anti_wind(),
            "bass_boost": self.get_bass_boost(),
            "channel": self.channel,
        }

File: src/headset.rs
//! Qualcomm GAIA Protocol client for Sennheiser Bluetooth Headphones (HDB 630, Momentum 4, etc.).
//! Communicates over native Linux Bluetooth RFCOMM sockets (`AF_BLUETOOTH`).

use std::fs;
use std::io;
use std::os::unix::io::RawFd;
use std::path::PathBuf;
use std::process::Command;
use std::time::{Duration, Instant};
use serde::{Deserialize, Serialize};

use crate::constants::*;

pub const BTPROTO_RFCOMM: libc::c_int = 3;

#[repr(C)]
#[derive(Debug, Copy, Clone)]
pub struct SockaddrRc {
    pub rc_family: libc::sa_family_t,
    pub rc_bdaddr: [u8; 6],
    pub rc_channel: u8,
    pub _padding: u8,
}

pub fn parse_mac_to_bdaddr(mac: &str) -> Option<[u8; 6]> {
    let parts: Vec<&str> = mac.split(':').collect();
    if parts.len() != 6 {
        return None;
    }
    let mut bytes = [0u8; 6];
    for (i, part) in parts.iter().enumerate() {
        bytes[5 - i] = u8::from_str_radix(part, 16).ok()?;
    }
    Some(bytes)
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PairedDevice {
    pub mac: String,
    pub name: String,
    pub recommended: bool,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct HeadsetConfig {
    pub headset_mac: Option<String>,
    pub cached_channel: Option<u8>,
}

pub fn get_config_path() -> PathBuf {
    let home = std::env::var("HOME").unwrap_or_else(|_| ".".to_string());
    PathBuf::from(home).join(".config").join("btd700").join("config.json")
}

pub fn load_config() -> HeadsetConfig {
    let path = get_config_path();
    if let Ok(data) = fs::read_to_string(&path) {
        if let Ok(cfg) = serde_json::from_str(&data) {
            return cfg;
        }
    }
    HeadsetConfig::default()
}

pub fn save_config(cfg: &HeadsetConfig) {
    let path = get_config_path();
    if let Some(parent) = path.parent() {
        let _ = fs::create_dir_all(parent);
    }
    if let Ok(json) = serde_json::to_string_pretty(cfg) {
        let _ = fs::write(path, json);
    }
}

pub fn find_paired_headsets() -> Vec<PairedDevice> {
    let mut results = Vec::new();
    if let Ok(output) = Command::new("bluetoothctl").args(["--timeout", "2", "devices"]).output() {
        let stdout = String::from_utf8_lossy(&output.stdout);
        for line in stdout.lines() {
            let parts: Vec<&str> = line.split_whitespace().collect();
            if parts.len() >= 3 && parts[0] == "Device" {
                let mac = parts[1].to_uppercase();
                let name = parts[2..].join(" ");
                let lower = name.to_lowercase();
                let rec = lower.contains("hdb")
                    || lower.contains("630")
                    || lower.contains("sennheiser")
                    || lower.contains("momentum")
                    || lower.contains("accentum")
                    || lower.contains("hd 450")
                    || lower.contains("hd 350");
                results.push(PairedDevice {
                    mac,
                    name,
                    recommended: rec,
                });
            }
        }
    }
    results.sort_by_key(|d| !d.recommended);
    results
}

pub fn get_default_mac() -> Option<String> {
    let cfg = load_config();
    if let Some(mac) = cfg.headset_mac {
        if !mac.is_empty() {
            return Some(mac.to_uppercase());
        }
    }
    let devs = find_paired_headsets();
    for d in &devs {
        if d.recommended {
            return Some(d.mac.clone());
        }
    }
    devs.first().map(|d| d.mac.clone())
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct HeadsetState {
    pub mac: String,
    pub connected: bool,
    pub anc_enabled: bool,
    pub anc_strength: i32,
    pub transparency: i32,
    pub adaptive_anc: bool,
    pub anti_wind: String,
    pub bass_boost: bool,
    pub channel: Option<u8>,
}

pub struct GaiaProtocol;

impl GaiaProtocol {
    pub fn build(command: u16, payload: &[u8]) -> Vec<u8> {
        let mut pkt = Vec::with_capacity(8 + payload.len());
        pkt.extend_from_slice(&GAIA_MAGIC);
        pkt.extend_from_slice(&(payload.len() as u16).to_be_bytes());
        pkt.extend_from_slice(&VENDOR_SENNHEISER.to_be_bytes());
        pkt.extend_from_slice(&command.to_be_bytes());
        pkt.extend_from_slice(payload);
        pkt
    }

    pub fn parse_many(buffer: &[u8]) -> (Vec<(u16, u16, Vec<u8>)>, usize) {
        let mut packets = Vec::new();
        let mut offset = 0;

        while buffer.len().saturating_sub(offset) >= 8 {
            if buffer[offset..offset + 2] != GAIA_MAGIC {
                if let Some(pos) = buffer[offset + 1..]
                    .windows(2)
                    .position(|w| w == GAIA_MAGIC)
                {
                    offset += 1 + pos;
                    continue;
                } else {
                    return (packets, buffer.len());
                }
            }

            let payload_len = u16::from_be_bytes([buffer[offset + 2], buffer[offset + 3]]) as usize;
            let total = 8 + payload_len;
            if buffer.len() - offset < total {
                break;
            }

            let vendor = u16::from_be_bytes([buffer[offset + 4], buffer[offset + 5]]);
            let command = u16::from_be_bytes([buffer[offset + 6], buffer[offset + 7]]);
            let payload = buffer[offset + 8..offset + total].to_vec();
            packets.push((vendor, command, payload));
            offset += total;
        }

        (packets, offset)
    }
}

pub struct SennheiserHeadset {
    pub mac: String,
    pub channel: Option<u8>,
    sock: Option<RawFd>,
    buffer: Vec<u8>,
}

impl SennheiserHeadset {
    pub fn new(mac: Option<String>) -> Self {
        let target_mac = mac.or_else(get_default_mac).unwrap_or_default().to_uppercase();
        Self {
            mac: target_mac,
            channel: None,
            sock: None,
            buffer: Vec::new(),
        }
    }

    pub fn is_connected(&self) -> bool {
        self.sock.is_some()
    }

    pub fn close(&mut self) {
        if let Some(fd) = self.sock.take() {
            unsafe { libc::close(fd) };
        }
        self.channel = None;
        self.buffer.clear();
    }

    fn connect_channel(&mut self, channel: u8, timeout: Duration) -> io::Result<()> {
        self.close();
        let bdaddr = parse_mac_to_bdaddr(&self.mac)
            .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidInput, "Invalid MAC"))?;

        let fd = unsafe {
            libc::socket(
                libc::AF_BLUETOOTH,
                libc::SOCK_STREAM,
                BTPROTO_RFCOMM,
            )
        };
        if fd < 0 {
            return Err(io::Error::last_os_error());
        }

        // Set nonblocking to handle connect timeout
        unsafe {
            let flags = libc::fcntl(fd, libc::F_GETFL, 0);
            libc::fcntl(fd, libc::F_SETFL, flags | libc::O_NONBLOCK);
        }

        let addr = SockaddrRc {
            rc_family: libc::AF_BLUETOOTH as libc::sa_family_t,
            rc_bdaddr: bdaddr,
            rc_channel: channel,
            _padding: 0,
        };

        let ret = unsafe {
            libc::connect(
                fd,
                &addr as *const _ as *const libc::sockaddr,
                std::mem::size_of::<SockaddrRc>() as libc::socklen_t,
            )
        };

        if ret < 0 {
            let err = io::Error::last_os_error();
            if err.raw_os_error() != Some(libc::EINPROGRESS) {
                unsafe { libc::close(fd) };
                return Err(err);
            }

            // Wait with poll
            let mut pfd = libc::pollfd {
                fd,
                events: libc::POLLOUT,
                revents: 0,
            };
            let poll_ret = unsafe { libc::poll(&mut pfd, 1, timeout.as_millis() as libc::c_int) };
            if poll_ret <= 0 {
                unsafe { libc::close(fd) };
                return Err(io::Error::new(io::ErrorKind::TimedOut, "Connect timeout"));
            }

            let mut sock_err: libc::c_int = 0;
            let mut len = std::mem::size_of::<libc::c_int>() as libc::socklen_t;
            unsafe {
                libc::getsockopt(
                    fd,
                    libc::SOL_SOCKET,
                    libc::SO_ERROR,
                    &mut sock_err as *mut _ as *mut libc::c_void,
                    &mut len,
                );
            }
            if sock_err != 0 {
                unsafe { libc::close(fd) };
                return Err(io::Error::from_raw_os_error(sock_err));
            }
        }

        self.sock = Some(fd);
        self.channel = Some(channel);
        Ok(())
    }

    fn exchange(&mut self, command: u16, payload: &[u8], timeout: Duration) -> io::Result<(u16, u16, Vec<u8>)> {
        let fd = self.sock.ok_or_else(|| io::Error::new(io::ErrorKind::NotConnected, "Not connected"))?;
        let packet = GaiaProtocol::build(command, payload);

        let n = unsafe {
            libc::send(
                fd,
                packet.as_ptr() as *const libc::c_void,
                packet.len(),
                0,
            )
        };
        if n < 0 {
            return Err(io::Error::last_os_error());
        }

        let expected_resp = command | 0x0100;
        let expected_err = command | 0x0180;
        let start = Instant::now();

        while start.elapsed() < timeout {
            let (packets, consumed) = GaiaProtocol::parse_many(&self.buffer);
            self.buffer.drain(..consumed);

            for (v, c, p) in packets {
                if c == expected_resp || c == expected_err {
                    return Ok((v, c, p));
                }
            }

            let rem = timeout.saturating_sub(start.elapsed()).as_millis().max(50) as libc::c_int;
            let mut pfd = libc::pollfd {
                fd,
                events: libc::POLLIN,
                revents: 0,
            };
            let poll_ret = unsafe { libc::poll(&mut pfd, 1, rem) };
            if poll_ret > 0 && (pfd.revents & libc::POLLIN) != 0 {
                let mut buf = [0u8; 1024];
                let rd = unsafe {
                    libc::recv(
                        fd,
                        buf.as_mut_ptr() as *mut libc::c_void,
                        buf.len(),
                        0,
                    )
                };
                if rd <= 0 {
                    return Err(io::Error::new(io::ErrorKind::ConnectionReset, "RFCOMM connection closed"));
                }
                self.buffer.extend_from_slice(&buf[..rd as usize]);
            }
        }

        Err(io::Error::new(io::ErrorKind::TimedOut, "Command timed out"))
    }

    pub fn connect(&mut self, timeout: Duration) -> io::Result<u8> {
        if self.mac.is_empty() {
            return Err(io::Error::new(
                io::ErrorKind::NotFound,
                "No Sennheiser headset MAC configured or found in paired devices",
            ));
        }

        let mut cfg = load_config();
        let mut channels = COMMON_RFCOMM_CHANNELS.to_vec();
        if let Some(cached) = cfg.cached_channel {
            if let Some(pos) = channels.iter().position(|&c| c == cached) {
                channels.remove(pos);
                channels.insert(0, cached);
            }
        }

        for &ch in &channels {
            if self.connect_channel(ch, timeout).is_ok() {
                if let Ok((v, c, _)) = self.exchange(GAIA_CMD_GET_ANC_STATUS, &[], Duration::from_millis(1500)) {
                    if v == VENDOR_SENNHEISER && c == (GAIA_CMD_GET_ANC_STATUS | 0x0100) {
                        cfg.headset_mac = Some(self.mac.clone());
                        cfg.cached_channel = Some(ch);
                        save_config(&cfg);
                        return Ok(ch);
                    }
                }
                self.close();
            }
        }

        Err(io::Error::new(
            io::ErrorKind::ConnectionRefused,
            format!("Failed to connect to Sennheiser headset ({}) over RFCOMM channels", self.mac),
        ))
    }

    pub fn get_anc_status(&mut self) -> bool {
        if let Ok((v, _, p)) = self.exchange(GAIA_CMD_GET_ANC_STATUS, &[], Duration::from_millis(1500)) {
            if v == VENDOR_SENNHEISER && !p.is_empty() {
                return p[0] != 0;
            }
        }
        false
    }

    pub fn set_anc_status(&mut self, enabled: bool) -> bool {
        let payload = [if enabled { 1 } else { 0 }];
        if let Ok((v, _, _)) = self.exchange(GAIA_CMD_SET_ANC_STATUS, &payload, Duration::from_millis(1500)) {
            return v == VENDOR_SENNHEISER;
        }
        false
    }

    pub fn toggle_anc(&mut self) -> bool {
        let current = self.get_anc_status();
        let new_state = !current;
        self.set_anc_status(new_state);
        new_state
    }

    pub fn get_transparency(&mut self) -> i32 {
        if let Ok((v, _, p)) = self.exchange(GAIA_CMD_GET_TRANSPARENCY, &[], Duration::from_millis(1500)) {
            if v == VENDOR_SENNHEISER && !p.is_empty() {
                return p[0] as i32;
            }
        }
        0
    }

    pub fn set_transparency(&mut self, level: i32) -> bool {
        let lvl = level.clamp(0, 100) as u8;
        if let Ok((v, _, _)) = self.exchange(GAIA_CMD_SET_TRANSPARENCY, &[lvl], Duration::from_millis(1500)) {
            return v == VENDOR_SENNHEISER;
        }
        false
    }

    pub fn get_anc_strength(&mut self) -> i32 {
        if !self.get_anc_status() {
            return 0;
        }
        100 - self.get_transparency()
    }

    pub fn set_anc_strength(&mut self, strength: i32) -> bool {
        let s = strength.clamp(0, 100);
        if s > 0 {
            self.set_anc_status(true);
        }
        let trans = 100 - s;
        self.set_transparency(trans)
    }

    pub fn get_adaptive_anc(&mut self) -> bool {
        if let Ok((v, _, p)) = self.exchange(GAIA_CMD_GET_ADAPTIVE_ANC, &[], Duration::from_millis(1000)) {
            if v == VENDOR_SENNHEISER && !p.is_empty() {
                return p[0] != 0;
            }
        }
        false
    }

    pub fn set_adaptive_anc(&mut self, enabled: bool) -> bool {
        let payload = [if enabled { 1 } else { 0 }];
        if let Ok((v, _, _)) = self.exchange(GAIA_CMD_SET_ADAPTIVE_ANC, &payload, Duration::from_millis(1000)) {
            return v == VENDOR_SENNHEISER;
        }
        false
    }

    pub fn get_anti_wind(&mut self) -> String {
        if let Ok((v, _, p)) = self.exchange(GAIA_CMD_GET_ANTI_WIND, &[], Duration::from_millis(1000)) {
            if v == VENDOR_SENNHEISER && !p.is_empty() {
                return match p[0] {
                    1 => "auto",
                    2 => "max",
                    _ => "off",
                }
                .to_string();
            }
        }
        "off".to_string()
    }

    pub fn set_anti_wind(&mut self, mode: &str) -> bool {
        let val = match mode.to_lowercase().as_str() {
            "auto" => 1u8,
            "max" => 2u8,
            _ => 0u8,
        };
        if let Ok((v, _, _)) = self.exchange(GAIA_CMD_SET_ANTI_WIND, &[val], Duration::from_millis(1000)) {
            return v == VENDOR_SENNHEISER;
        }
        false
    }

    pub fn get_bass_boost(&mut self) -> bool {
        if let Ok((v, _, p)) = self.exchange(GAIA_CMD_GET_BASS_BOOST, &[], Duration::from_millis(1500)) {
            if v == VENDOR_SENNHEISER && !p.is_empty() {
                return p[0] != 0;
            }
        }
        false
    }

    pub fn set_bass_boost(&mut self, enabled: bool) -> bool {
        let payload = [if enabled { 1 } else { 0 }];
        if let Ok((v, _, _)) = self.exchange(GAIA_CMD_SET_BASS_BOOST, &payload, Duration::from_millis(1500)) {
            return v == VENDOR_SENNHEISER;
        }
        false
    }

    pub fn get_state(&mut self) -> HeadsetState {
        let anc = self.get_anc_status();
        let trans = self.get_transparency();
        HeadsetState {
            mac: self.mac.clone(),
            connected: self.is_connected(),
            anc_enabled: anc,
            anc_strength: if anc { 100 - trans } else { 0 },
            transparency: trans,
            adaptive_anc: self.get_adaptive_anc(),
            anti_wind: self.get_anti_wind(),
            bass_boost: self.get_bass_boost(),
            channel: self.channel,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_mac_to_bdaddr() {
        let mac = "00:11:22:33:44:55";
        let addr = parse_mac_to_bdaddr(mac);
        assert!(addr.is_some());
        // Little endian reverse order: 55:44:33:22:11:00
        assert_eq!(addr.unwrap(), [0x55, 0x44, 0x33, 0x22, 0x11, 0x00]);

        assert_eq!(parse_mac_to_bdaddr("invalid"), None);
        assert_eq!(parse_mac_to_bdaddr("00:11:22:33:44"), None);
        assert_eq!(parse_mac_to_bdaddr("00:11:22:33:44:55:66"), None);
        assert_eq!(parse_mac_to_bdaddr("00:11:22:33:44:ZZ"), None);
    }

    #[test]
    fn test_gaia_build() {
        let cmd = GAIA_CMD_SET_ANC_STATUS;
        let payload = [0x01];
        let pkt = GaiaProtocol::build(cmd, &payload);

        assert_eq!(&pkt[0..2], &GAIA_MAGIC);
        assert_eq!(u16::from_be_bytes([pkt[2], pkt[3]]), 1); // length
        assert_eq!(u16::from_be_bytes([pkt[4], pkt[5]]), VENDOR_SENNHEISER);
        assert_eq!(u16::from_be_bytes([pkt[6], pkt[7]]), cmd);
        assert_eq!(&pkt[8..], &[0x01]);
    }

    #[test]
    fn test_gaia_parse_many_single() {
        let pkt = GaiaProtocol::build(GAIA_CMD_GET_ANC_STATUS, &[0x01]);
        let (packets, consumed) = GaiaProtocol::parse_many(&pkt);
        assert_eq!(packets.len(), 1);
        assert_eq!(consumed, pkt.len());
        assert_eq!(packets[0].0, VENDOR_SENNHEISER);
        assert_eq!(packets[0].1, GAIA_CMD_GET_ANC_STATUS);
        assert_eq!(packets[0].2, vec![0x01]);
    }

    #[test]
    fn test_gaia_parse_many_multiple() {
        let mut buffer = Vec::new();
        buffer.extend_from_slice(&GaiaProtocol::build(0x1001, &[1, 2]));
        buffer.extend_from_slice(&GaiaProtocol::build(0x1002, &[3, 4, 5]));

        let (packets, consumed) = GaiaProtocol::parse_many(&buffer);
        assert_eq!(packets.len(), 2);
        assert_eq!(consumed, buffer.len());
        assert_eq!(packets[0].1, 0x1001);
        assert_eq!(packets[0].2, vec![1, 2]);
        assert_eq!(packets[1].1, 0x1002);
        assert_eq!(packets[1].2, vec![3, 4, 5]);
    }

    #[test]
    fn test_gaia_parse_many_with_garbage_prefix() {
        let mut buffer = vec![0x00, 0xAA, 0xBB];
        buffer.extend_from_slice(&GaiaProtocol::build(0x1A04, &[0x00]));

        let (packets, consumed) = GaiaProtocol::parse_many(&buffer);
        assert_eq!(packets.len(), 1);
        assert_eq!(consumed, buffer.len());
        assert_eq!(packets[0].1, 0x1A04);
        assert_eq!(packets[0].2, vec![0x00]);
    }

    #[test]
    fn test_headset_config_serde() {
        let cfg = HeadsetConfig {
            headset_mac: Some("AA:BB:CC:DD:EE:FF".to_string()),
            cached_channel: Some(2),
        };
        let json = serde_json::to_string(&cfg).unwrap();
        let loaded: HeadsetConfig = serde_json::from_str(&json).unwrap();
        assert_eq!(loaded.headset_mac.as_deref(), Some("AA:BB:CC:DD:EE:FF"));
        assert_eq!(loaded.cached_channel, Some(2));
    }
}

